cmd/mvpage/mover.go

```diff
@@ -0,0 +1,188 @@
+// Copyright 2019 Istio Authors
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"gopkg.in/yaml.v2"
+)
+
+type mover struct {
+	// set when creating the object
+	config hugoConfig
+
+	// work space for individual move operations
+	srcFile      string
+	srcLink      string
+	dstFile      string
+	dstLink      string
+	srcLinkRegex *regexp.Regexp
+}
+
+// The regex we use to isolate front matter in a merkdown file
+var frontMatterRE = regexp.MustCompile("(?ms)^---$(.*)^---$")
+
+func newMover(config hugoConfig) *mover {
+	return &mover{
+		config: config,
+	}
+}
+
+// Does all the work needed to move from src to dst
+func (m *mover) move(srcFile string, dstFile string) error {
+	var err error
+
+	if m.srcFile, m.srcLink, err = m.validate(srcFile); err != nil {
+		return err
+	}
+
+	if m.dstFile, m.dstLink, err = m.validate(dstFile); err != nil {
+		return err
+	}
+
+	if m.srcLinkRegex, err = regexp.Compile(regexp.QuoteMeta("(" + m.srcLink)); err != nil {
```

Comment:
should you match the closing ')' too?  otherwise you might replace a prefix of a different link?

Reply:
I wanted to avoid problems with trailing fragments.  That  is (link#fragment). I guess I could incorporate optional fragments into the regex.